[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #118584
* #118298
* #118223
* #118198
* #118188
* #118136
* #118118
* #118017
* #118004
* #117975
* #117973
* #117955
* #117950
* #117881
* #117877
* __->__ #117867
* #117814
* #118525

This PR adds the initial `FSDPParamGroup` and `FSDPParam` classes, and it focuses on the `ParamModuleInfo` data structure.

- `ParamModuleInfo` has the info needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.
    ```
    # Shared parameter
    lin1.weight = lin2.weight

    # Shared module
    mlp.lin1 = mlp.lin2
    ```
- In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

With this PR, we see the general system architecture:
- 1 `module` : 1 `fully_shard`
- 1 `fully_shard` : 1 `FSDPParamGroup`
- 1 `FSDPParamGroup` : k `FSDPParam`


cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

Commit:
[no ci][FSDP2] Added initial `FSDPParamGroup`

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup` and `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` class and focuses on `ParamModuleInfo`. `ParamModuleInfo` is a data structure with the information needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.

```
# Shared parameter
lin1.weight = lin2.weight

# Shared module
mlp.lin1 = mlp.lin2
```

In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` class and focuses on `ParamModuleInfo`. `ParamModuleInfo` is a data structure with the information needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.

```
# Shared parameter
lin1.weight = lin2.weight

# Shared module
mlp.lin1 = mlp.lin2
```

In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` and `FSDPParam` classes, and it focuses on the `ParamModuleInfo` data structure.

- `ParamModuleInfo` has the info needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.
    ```
    # Shared parameter
    lin1.weight = lin2.weight

    # Shared module
    mlp.lin1 = mlp.lin2
    ```
- In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

With this PR, we see the general system architecture:
- 1 `module` : 1 `fully_shard`
- 1 `fully_shard` : 1 `FSDPParamGroup`
- 1 `FSDPParamGroup` : k `FSDPParam`


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

weifengpy commented on torch/distributed/_composable/fsdp/_fsdp_common.py:
```diff
@@ -53,6 +53,22 @@ def __post_init__(self):
 
 
+@dataclass
+class ParamModuleInfo:
```
Nit: Is ``ParamModuleInfo`` used in used in ``FSDPParam`` only ? wonder if we should put it in ``_fsdp_common.py`` or ``_flat_param.py``

awgu replied:
This is a good point! Since `ParamModuleInfo` is used in both `FSDPParam` and `FSDPParamGroup`, I had `ParamModuleInfo` in `_fsdp_common.py`. However, I realize now that this is not entirely consistent (e.g. `ShardedState` is used in both but is in `_fsdp_param.py`), and this inconsistency is a consequence of how the code evolved over time.

Do you prefer these to be in `_fsdp_param.py`?

If so, I might prefer to move it in a separate PR because changing the imports causes so many rebase conflicts 😢 , but I can do it here if desired too.

weifengpy replied:
separate PR works for me. other files import from `_fsdp_param.py`since it ultimately serves ``FSDPParam``

awgu replied:
To be good, I am going to try to fix this directly here and handle the rebase conflicts 😅

weifengpy commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -0,0 +1,50 @@
+) -> List[ParamModuleInfo]:
+    params_set = set(params)
+    param_to_module_info: Dict[nn.Parameter, ParamModuleInfo] = {}
+    # Do not remove duplicates to find parameters from shared modules
```
Nit: I misunderstand and thought shared modules was referring to `Transformer` and `Block1` since they share `param` `Transformer(Block(param))`. But actually it's referring to `Block1` and `Block2` if Block1.param == Block2.param

maybe considering this docstring
```
\# Different modules may refer to the same parameter instance
\# Do not remove duplicates to find parameters shared by different modules
```

weifengpy commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -53,6 +54,8 @@ def fully_shard(
     _move_states_to_device(params, buffers, device, mesh_info)
+    if params:
+        state._fsdp_param_group = FSDPParamGroup(params, module, mesh_info, device)
 
```
non-blocking question for buffers: are we not sharding them because they usually consume small memories? even in finetuning, it's ``nn.parameters`` (requries_grad=False) instead of buffers?

awgu replied:
Yep!. Buffers themselves usually are small, and moreover, since they are not trainable, they never have optimizer state or gradients.

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` and `FSDPParam` classes, and it focuses on the `ParamModuleInfo` data structure.

- `ParamModuleInfo` has the info needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.
    ```
    # Shared parameter
    lin1.weight = lin2.weight

    # Shared module
    mlp.lin1 = mlp.lin2
    ```
- In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

With this PR, we see the general system architecture:
- 1 `module` : 1 `fully_shard`
- 1 `fully_shard` : 1 `FSDPParamGroup`
- 1 `FSDPParamGroup` : k `FSDPParam`


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` and `FSDPParam` classes, and it focuses on the `ParamModuleInfo` data structure.

- `ParamModuleInfo` has the info needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.
    ```
    # Shared parameter
    lin1.weight = lin2.weight

    # Shared module
    mlp.lin1 = mlp.lin2
    ```
- In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

With this PR, we see the general system architecture:
- 1 `module` : 1 `fully_shard`
- 1 `fully_shard` : 1 `FSDPParamGroup`
- 1 `FSDPParamGroup` : k `FSDPParam`


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` and `FSDPParam` classes, and it focuses on the `ParamModuleInfo` data structure.

- `ParamModuleInfo` has the info needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.
    ```
    # Shared parameter
    lin1.weight = lin2.weight

    # Shared module
    mlp.lin1 = mlp.lin2
    ```
- In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

With this PR, we see the general system architecture:
- 1 `module` : 1 `fully_shard`
- 1 `fully_shard` : 1 `FSDPParamGroup`
- 1 `FSDPParamGroup` : k `FSDPParam`


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` and `FSDPParam` classes, and it focuses on the `ParamModuleInfo` data structure.

- `ParamModuleInfo` has the info needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.
    ```
    # Shared parameter
    lin1.weight = lin2.weight

    # Shared module
    mlp.lin1 = mlp.lin2
    ```
- In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

With this PR, we see the general system architecture:
- 1 `module` : 1 `fully_shard`
- 1 `fully_shard` : 1 `FSDPParamGroup`
- 1 `FSDPParamGroup` : k `FSDPParam`


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` and `FSDPParam` classes, and it focuses on the `ParamModuleInfo` data structure.

- `ParamModuleInfo` has the info needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.
    ```
    # Shared parameter
    lin1.weight = lin2.weight

    # Shared module
    mlp.lin1 = mlp.lin2
    ```
- In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

With this PR, we see the general system architecture:
- 1 `module` : 1 `fully_shard`
- 1 `fully_shard` : 1 `FSDPParamGroup`
- 1 `FSDPParamGroup` : k `FSDPParam`


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

wanchaol reviewed (approved):
lgtm! some comments and questions inlined.

awgu commented on torch/distributed/_composable/fsdp/fully_shard.py:
```diff
@@ -54,6 +55,8 @@ def fully_shard(
     managed_modules = _get_managed_modules(module)
     params, buffers = _get_managed_states(managed_modules)
     _move_states_to_device(params, buffers, device, mesh_info)
+    if params:
```
Correct!

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -0,0 +1,57 @@
+        for param_name, param in _named_parameters_with_duplicates(
+            submodule, recurse=False
+        ):
+            if param in params_set:
```
With the current code, it is possible if we have a nested `replicate` module, in which case the `replicate` module's parameters can be included in the `named_modules()` / `named_parameters()` traversal but not in `params_set` (which comes from `params` which has already filtered the `replicate` module's parameters).

In the future, if we allow for ignoring specific parameters, then that could be another source where `param not in params_set`.

awgu commented on torch/distributed/_composable/fsdp/_fsdp_param_group.py:
```diff
@@ -0,0 +1,57 @@
+                    param_to_module_info[param] = ParamModuleInfo(submodule, param_name)
+                else:
+                    param_to_module_info[param].shared_modules.append(submodule)
+                    param_to_module_info[param].shared_param_names.append(param_name)
```
I think a `Dict[str, nn.Module]` may not work for this usage. We want the unprefixed parameter name (e.g. `weight`, not `lin.weight`) so that we can `setattr(module, param_name, tensor)` with the unprefixed `param_name`.

However, we could have a case like: `lin1.weight = lin2.weight = lin3.weight`, in which case all of the `param_name`s are `"weight"` and `"weight"` will no longer be a unique key. Alternatively, we could `.split('.')[-1]` each time we use the parameter name, but the current 2-list approach might be more direct.

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` and `FSDPParam` classes, and it focuses on the `ParamModuleInfo` data structure.

- `ParamModuleInfo` has the info needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.
    ```
    # Shared parameter
    lin1.weight = lin2.weight

    # Shared module
    mlp.lin1 = mlp.lin2
    ```
- In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

With this PR, we see the general system architecture:
- 1 `module` : 1 `fully_shard`
- 1 `fully_shard` : 1 `FSDPParamGroup`
- 1 `FSDPParamGroup` : k `FSDPParam`


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

Commit:
Update on "[FSDP2] Added initial `FSDPParamGroup`, `FSDPParam`, `ParamModuleInfo`"


This PR adds the initial `FSDPParamGroup` and `FSDPParam` classes, and it focuses on the `ParamModuleInfo` data structure.

- `ParamModuleInfo` has the info needed to `setattr` a managed parameter, where it must account for shared parameters and shared modules.
    ```
    # Shared parameter
    lin1.weight = lin2.weight

    # Shared module
    mlp.lin1 = mlp.lin2
    ```
- In order for FSDP to find shared modules' parameters, we must use `remove_duplicate=False`. See https://github.com/pytorch/pytorch/pull/99448/ for the original context. Finding shared modules' parameters is not necessary for the `setattr` logic, but in case we need it in the future (like for existing FSDP's state dict), we include that info for now.

With this PR, we see the general system architecture:
- 1 `module` : 1 `fully_shard`
- 1 `fully_shard` : 1 `FSDPParamGroup`
- 1 `FSDPParamGroup` : k `FSDPParam`


cc mrshenli pritamdamania87 zhaojuanmao satgera rohan-varma gqchen aazzolini osalpekar jiayisuse H-Huang kwen2501 penguinwu fegin XilunWu wanchaol fduwjj wz337 tianyu-l wconstab yf225

[ghstack-poisoned]

awgu commented:
@pytorchbot merge

awgu commented:
Manually triggered test run shows all green tests: https://github.com/pytorch/pytorch/actions/runs/7716221399

awgu commented:
@pytorchbot merge -f "ci will not run; ran manually instead and verified green"
